gecko/glue/nsThread.cpp

```diff
@@ -463,8 +463,32 @@ nsThread::ThreadFunc(void* aArg)
 
   mozilla::IOInterposer::UnregisterCurrentThread();
 #else
-  self->mShutdownNow = false;
-  while (!self->mShutdownNow) {
+
+  // Under high load the thread might shut down before starting. So this first
+  // loop would spin forever. Hence the test on mShutdownNow. This loop is the
+  // GeckoMedia equivalent of the message loop deactivated above.
+  if (!self->mShutdownNow) {
+    while (true) {
+      if (NS_ProcessNextEvent(self, false))
+        continue;
+
+      NS_ProcessNextEvent(self, true);
+
+      MonitorAutoLock mon(self->mShutdownMonitor);
+      mon.Wait(10);
+      if (self->mShutdownNow)
```

Comment:
I think we're better off just relying on nsThread::ShuttingDown(), as that causes nsThread::ProcessNextEvent() and thus NS_ProcessNextEvent() to exit already, where as it's not clear this flag does. Using the existing mechanism  will likely be more resilient to changes in the upstream code in future too. I'll make a PR for that.